### ISSUE TITLE
[cloud_firestore] create document only if it doesn't already exists

### ISSUE THREAD

**Issue by kroikie** [FirebaseExtended/flutterfire#1203]
## Use case

I'm trying to create a document in Firestore with a predictable id. I want to make sure that the document is only created once, and never overwritten. In our cloud functions, we are using the JavaScript `create` function for this:

```javascript
const admin = require('firebase-admin')

// ...

const ref = admin.firestore().collection('foobar').doc(id)

await ref.create({ foo: 'bar' })
```

It seems like the `create` function is missing from the Flutter version of the API though, only `set` and `update` is present.

## Proposal

It would be great if we could add `createData` in addition to `setData` and `updateData`:

```dart
import 'package:cloud_firestore/cloud_firestore.dart';

// ...

final ref = Firestore.instance.collection('foobar').document(id);

await ref.createData({ 'foo': 'bar' })
```

**Comment by kroikie** [comment]
@LinusU

The issue at https://github.com/flutter/flutter/issues/38806 has been closed and moved here. Future collaboration on this issue will be done here.

**Comment by Salakar** [comment]
Hey, as part of our on-going work for #2582, this has been resolved in our Firebase Firestore rework (#2913) (you can now use `SetOptions` to achieve the desired behaviour above) - which has now been merged into master. We'll look at publishing some prereleases in the next few days. Thank you
